materialize-sqlite/sqlite.go

```diff
@@ -0,0 +1,357 @@
+
+	pf "github.com/estuary/flow/go/protocols/flow"
+	pm "github.com/estuary/flow/go/protocols/materialize"
+	sql "github.com/estuary/connectors/materialize-sql"
+	boilerplate "github.com/estuary/connectors/materialize-boilerplate"
+	_ "github.com/mattn/go-sqlite3" // Import for register side-effects.
+	log "github.com/sirupsen/logrus"
+)
+
+type config struct {
+	path string
+}
+func (c config) Validate() error {
+	return nil
+}
+
+type tableConfig struct {
+	Table string `json:"table"`
+}
+
+func (c tableConfig) Validate() error {
+	if c.Table == "" {
+		return fmt.Errorf("expected SQLite database configuration `table`")
+	}
+	return nil
+}
+
+func (c tableConfig) GetAdditionalSql() string {
+	return ""
+}
+
+func (c tableConfig) Path() sql.TablePath {
+	return []string{c.Table}
+}
+
+func (c tableConfig) DeltaUpdates() bool {
+	return false // SQLite doesn't support delta updates.
+}
+
+func newTableConfig(ep *sql.Endpoint) sql.Resource {
+	return &tableConfig{}
+}
+
+// NewSQLiteDriver creates a new Driver for sqlite.
+func NewSQLiteDriver() *sql.Driver {
+	return &sql.Driver{
+		DocumentationURL: "https://docs.estuary.dev/#FIXME",
+		EndpointSpecType: new(config),
+		ResourceSpecType: new(tableConfig),
+		NewEndpoint: func(ctx context.Context, _ json.RawMessage) (*sql.Endpoint, error) {
+			var path = "/tmp/sqlite.db"
+
+			// SQLite / go-sqlite3 is a bit fickle about raced opens of a newly created database,
+			// often returning "database is locked" errors. We can resolve by ensuring one sql.Open
+			// completes before the next starts. This is only required for SQLite, not other drivers.
+			sqliteOpenMu.Lock()
+			db, err := stdsql.Open("sqlite3", path)
+			if err == nil {
+				err = db.PingContext(ctx)
+			}
+			sqliteOpenMu.Unlock()
+
+			if err != nil {
+				return nil, fmt.Errorf("opening SQLite database %q: %w", path, err)
+			}
+
+			return &sql.Endpoint{
+				Config: config{path: path},
+				Dialect: sqliteDialect,
+				MetaSpecs: nil,
+				MetaCheckpoints: nil,
+				Client: client{path: path},
+				CreateTableTemplate: tplCreateTargetTable,
+				NewResource: newTableConfig,
+				NewTransactor: newTransactor,
+			}, nil
+		},
+	}
+}
+
+type client struct {
+	path string
+}
+
+// We don't use specs table for sqlite since it is ephemeral and won't be
+// persisted between ApplyUpsert and Transactions calls
+func (c client) FetchSpecAndVersion(ctx context.Context, specs sql.Table, materialization pf.Materialization) (specB64, version string, err error) {
+	return "", "", stdsql.ErrNoRows
+}
+
+func (c client) ExecStatements(ctx context.Context, statements []string) error {
+	return c.withDB(func(db *stdsql.DB) error { return sql.StdSQLExecStatements(ctx, db, statements) })
+}
+
+// We don't need a fence since there is only going to be a single sqlite
+// materialization instance writing to the file, and it's ephemeral
+func (c client) InstallFence(ctx context.Context, checkpoints sql.Table, fence sql.Fence) (sql.Fence, error) {
```

Comment:
This is interesting: Since the driver won't supply a checkpoint to the runtime, the runtime will always resume from the recovery log checkpoint. The alternative would be re-starting the materialization from scratch (by returning `pm.ExplicitZeroCheckpoint`). Starting from scratch every time would probably not be great since it could result in huge data transfers every time the connector restarts. Starting from the recovery log checkpoint might be a little surprising to users. I think thats still the best option though since this is meant for internal testing and demos.

Reply:
Yeah this is a bit ambiguous at the moment, we can play with the current approach and if we have a reason to use a zero checkpoint, we'll go for that instead.